### PR TITLE
Internal lints: usage_of_qualified_ty & ty_pass_by_reference

### LINKED ISSUE #59952: Enforce import / type code styles throughout the compiler.
We've had a few large-scale cleanup PRs before, but it doesn't always stick.
This is a quick example of what I'm referring to (as the agreed-upon style):
```rust
use rustc::ty::{Ty, TyCtxt};
fn foo(tcx: TyCtxt<'_, '_, 'tcx>, ty: Ty<'tcx>) {...}
```
`TyCtxt` and `Ty` are referred to unqualified (without `ty::`) and passed by value.
(Note that types can be behind `&` due to a trait, but that should be *the only* exception to the rule)

Currently both `&TyCtxt` and `&Ty<'tcx>` show up in the codebase (although in small quantities), but I only accidentally noticed and I won't be replacing them any time soon (someone else can, if they want to, but I'm more interested in how we keep this from happening - maybe with `clippy`?).

cc @rust-lang/compiler

### PR DESCRIPTION
**Pull request by flip1995**
Closes #59952 

Implements internal lints:

- `USAGE_OF_QUALIFIED_TY`
- `TY_PASS_BY_REFERENCE`

r? @oli-obk

### CODE REVIEW

File: src/librustc/lint/internal.rs [hunk 1]
```diff
@@ -107,3 +179,44 @@ fn lint_ty_kind_usage(cx: &LateContext<'_, '_>, segment: &PathSegment) -> bool {
 
     false
 }
+
+fn is_ty_or_ty_ctxt(cx: &LateContext<'_, '_>, ty: &Ty) -> Option<String> {
+    match &ty.node {
+        TyKind::Path(qpath) => {
+            if let QPath::Resolved(_, path) = qpath {
+                if let Some(did) = path.def.opt_def_id() {
```
Comment (oli-obk) [inline src/librustc/lint/internal.rs]: use the `?` operator to get rid of one level of nesting here

**Comment by oli-obk** [comment]
needs a rebase, r=me with that

@bors delegate+

**Comment by flip1995** [comment]
Also removed the unused argument from the `allow_two_phase_borrow` function. cc #60347

@bors r=oli-obk

File: src/librustc/mir/visit.rs [hunk 2]
```diff
@@ -198,7 +198,7 @@ macro_rules! make_mir_visitor {
             }
 
             fn visit_ty(&mut self,
-                        ty: & $($mutability)? Ty<'tcx>,
+                        ty: $(& $mutability)? Ty<'tcx>,
```
Comment (eddyb) [inline src/librustc/mir/visit.rs]: Wow, that's a neat trick! cc @nikomatsakis how did we not think of this?

Comment (eddyb) [reply src/librustc/mir/visit.rs]: `{visit,super}_{local,region,const}` (and maybe a few others?) should probably get the same treatment.

Comment (nikomatsakis) [reply src/librustc/mir/visit.rs]: heh, a good question.

**Pull request merged**